src/resolve/index.ts

```diff
@@ -6,17 +6,73 @@ import {Got} from '../network/got'
 import {Package} from '../types'
 import {LoggedPkg} from 'pnpm-logger'
 
-export type ResolveResult = {
+export type ResolutionBase = {
   id: string,
+}
+
+/**
+ * npm registry hosted package
+ */
+export type PackageResolution = ResolutionBase & {
+  type: 'package',
+  tarball: string,
+  shasum?: string,
   pkg?: Package,
+}
+
+/**
+ * tarball hosted remotely
+ */
+export type TarballResolution = ResolutionBase & {
+  type: 'tarball',
+  tarball: string,
+  shasum?: string,
+}
+
+/**
+ * tarball on a filesystem
+ */
+export type LocalTarballResolution = ResolutionBase & {
+  type: 'local-tarball',
+  tarball: string,
+  shasum?: string,
+}
+
+/**
+ * directory on a file system
+ */
+export type LocalDirectoryResolution = ResolutionBase & {
+  type: 'directory',
+  root: string,
+}
+
+/**
+ * directory on a file system, which should be linked
+ */
+export type LinkResolution = ResolutionBase & {
+  type: 'link',
+  root: string,
+}
+
+/**
+ * Git repository
+ */
+export type GitRepositoryResolution = ResolutionBase & {
+  type: 'git-repo',
+  // In case of some git repos we can resolve to the link to the tarball directly
   tarball?: string,
```

Comment:
I wonder if tarball should be in this type. Maybe git resolver can return either a GitRepositoryResolution or a TarballResolution